### PR TITLE
Amt provider interface

### LINKED ISSUE #131: Add support for iAMT
Is it feasible to add support for [iAMT](https://en.wikipedia.org/wiki/Intel_Active_Management_Technology) (used to manage kiosk like devices)? Or is that out-of-scope?

There is at least the the [MeshCentral application/library](https://github.com/Ylianst/MeshCentral) (written in node.js) that seems to have good support for it.

### PR DESCRIPTION
**Pull request by jacobweinstock**
## What does this PR implement/change/remove?

This PR adds the AMT provider.  Fixes #131 

### Checklist
- [X] Tests added
- [X] Similar commits squashed

### The HW vendor this change applies to (if applicable)
intel NUC with AMT support.

### What version of tooling - vendor specific or opensource does this change depend on (if applicable)
github.com/ammmze/go-amt
github.com/ammmze/wsman

## Description for changelog/release notes

```
Adds support for AMT out of band communication.
```

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -19,9 +22,21 @@ require (
...
+	github.com/satori/go.uuid v1.2.0 // indirect
+	golang.org/x/lint v0.0.0-20210508222113-6edffad5e616 // indirect
+	golang.org/x/mod v0.5.1 // indirect
 	golang.org/x/sys v0.0.0-20220520151302-bc2c85ada10a // indirect
+	golang.org/x/tools v0.1.8 // indirect
+	golang.org/x/xerrors v0.0.0-20200804184101-5ec99f83aff1 // indirect
 	gopkg.in/yaml.v3 v3.0.1 // indirect
 )
+
+replace github.com/ammmze/go-amt => github.com/jacobweinstock/go-amt v0.0.0-20221115190754-36c39f21b864
```
Comment (joelrebel) [inline go.mod:39]: Are there plans to upstream the changes in your fork?

File: providers/amt/amt.go [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+
+	"github.com/ammmze/go-amt"
+	"github.com/bmc-toolbox/bmclib/providers"
+	"github.com/go-logr/logr"
+	"github.com/jacobweinstock/registrar"
+)
+
+const (
+	// ProviderName for the provider AMT implementation
+	ProviderName = "AMT"
```
Comment (joelrebel) [inline providers/amt/amt.go:17]: Mind if we call this package and its `ProviderName` 'IntelAMT' instead, just so its a bit more indicative of the hardware vendor in the provider name.

**Review by joelrebel — CHANGES_REQUESTED** [review]
Thanks for these additions! I've suggested a change in the Provider name, apart from that, LGTM!

File: providers/amt/amt.go [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+	case "off":
+		if !on {
+			return true, nil
+		}
+		if err := c.client.PowerOff(ctx); err != nil {
+			return false, err
+		}
+		ok = true
+	case "cycle":
+		if !on {
```
Comment (rgl) [inline providers/amt/amt.go:164]: why is this needed? it seems [go-amt powerCycle already does it](https://github.com/jacobweinstock/go-amt/blob/main/power.go#L119).

File: providers/amt/amt.go (hunk 2 above)
Comment (jacobweinstock) [reply providers/amt/amt.go:17]: sure thing

File: go.mod (hunk 1 above)
Comment (jacobweinstock) [reply go.mod:39]: yes, will open a PR shortly.

File: providers/amt/amt.go [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+
+	return c
+}
+
+// Name of the provider
+func (c *Conn) Name() string {
+	return ProviderName
+}
+
+// Open a connection to a the BMC via AMT.
```
[the remark this answers is not in the document]
Comment (jacobweinstock) [reply providers/amt/amt.go:81]: ah yes. thanks

File: providers/amt/amt.go (hunk 3 above)
Comment (jacobweinstock) [reply providers/amt/amt.go:164]: ah, good catch. thanks. will update.
